[SPARK-11262][ML] Unit test for gradient, loss layers, memory management for multilayer perceptron

avulanov opened the pull request:
1.Implement LossFunction trait and implement squared error and cross entropy
loss with it
2.Implement unit test for gradient and loss
3.Implement InPlace trait and in-place layer evaluation
4.Refactor interface for ActivationFunction
5.Update of Layer and LayerModel interfaces
6.Fix random weights assignment
7.Implement memory allocation by MLP model instead of individual layers

These features decreased the memory usage and increased flexibility of
internal API.

avulanov committed:
Refactor Multilayer perceptron (MLP)

1.Implement LossFunction trait and implement squared error and cross entropy
loss with it
2.Implement unit test for gradient and loss
3.Implement InPlace trait and in-place layer evaluation
4.Refactor interface for ActivationFunction
5.Update of Layer and LayerModel interfaces
6.Fix random weights assignment
7.Implement memory allocation by MLP model instead of individual layers

These features decreased the memory usage and increased flexibility of
internal API.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44163 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44163/consoleFull)** for PR 9229 at commit [`d47310b`](https://github.com/apache/spark/commit/d47310b4a0e0229f96c2aded7c5fd9158ef26983).

SparkQA commented:
**[Test build #44163 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44163/consoleFull)** for PR 9229 at commit [`d47310b`](https://github.com/apache/spark/commit/d47310b4a0e0229f96c2aded7c5fd9158ef26983).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `trait LossFunction `\n  * `trait InPlace `\n  * `class SigmoidLayerWithSquaredError extends Layer `\n  * `class SigmoidLayerModelWithSquaredError extends FunctionalLayerModel(new SigmoidFunction)`\n  * `class SoftmaxLayerWithCrossEntropyLoss extends Layer `\n  * `class SoftmaxLayerModelWithCrossEntropyLoss extends LayerModel with LossFunction `\n

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44163/
Test FAILed.

avulanov commented:
@mengxr Could you take a look?

@feynmanliang These changes clash with some of your changes in https://github.com/apache/spark/pull/8648, sorry, because I simplified the interface of ActivationFunction to `Double => Double` and forced all computations to be in place. However, your suggestion of `UFunc` could be very helpful. Can you think of a way rewriting the `UniversalFunction` object with `UFunc` so that it does in-place operations? In any case, we have to merge your changes in `BreezeUtil.scala` either in the original PR or here and improve the scaladoc as you proposed.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -151,42 +160,39 @@ private[ann] object AffineLayerModel {
   def unroll(
-    weights: Vector,
-    position: Int,
+    weights: BDV[Double],
```
4 indent

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,117 @@
+/**
+ * Trait for loss function
+ */
+trait LossFunction {
```
Most of this file should be `private[ann]`

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -32,20 +32,33 @@ import org.apache.spark.util.random.XORShiftRandom
    */
-  def getInstance(weights: Vector, position: Int): LayerModel
-
+  def instance(weights: BDV[Double]): LayerModel
```
`Layer` is really implementing the Factory pattern for `LayerModel`. Should we consider renaming it to use the proper name?

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -111,32 +126,26 @@ private[ann] class AffineLayer(val numIn: Int, val numOut: Int) extends Layer {
+    // compute gradient of weights
+    val cumGradientOfWeights = new BDM[Double](w.rows, w.cols, cumGrad.data, cumGrad.offset)
+    BreezeUtil.dgemm(1.0 / input.cols, delta, input.t, 1.0, cumGradientOfWeights)
     if (ones == null || ones.length != delta.cols) ones = BDV.ones[Double](delta.cols)
```
Since `w` does not resize if `delta` changes, I don't think we need to worry about resizing `ones` here and can instead just make L130 lazy init `ones`

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -208,14 +214,18 @@ private[ann] object AffineLayerModel {
-    val bias = BDV.fill[Double](numOut){ (rand.nextDouble * 4.8 - 2.4) / numIn }
-    (weights, bias)
+  def randomWeights(numIn: Int, numOut: Int, weights: BDV[Double], random: Random):
+  (BDM[Double], BDV[Double]) = {
```
nit: I would prefer to line break at arguments rather than return type. Not sure if it's part of official style guide but it's what I've seen elsewhere

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,117 @@
+}
+
+trait InPlace {
+
```
No need for `{...}`

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -546,65 +446,83 @@ private[ml] object FeedForwardTopology {
+      var inputSize = data.rows
+      for (i <- 0 until layerModels.length) {
+        layerModels(i) match {
+          case inPlace: InPlace => outputs(i) = outputs(i - 1)
```
Since I don't see `InPlace` being mixed in anywhere except `LayerModel`, can we just replace the empty `InPlace` trait with a `inPlace : Boolean` flag defined in the `LayerModel` interface and reduce unnecessary abstraction?

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -546,65 +446,83 @@ private[ml] object FeedForwardTopology {
-    target: BDM[Double],
-    cumGradient: Vector,
-    realBatchSize: Int): Double = {
+  override def computeGradient(data: BDM[Double],
```
nit: Arguments should be 4-indented on new line

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -546,65 +446,83 @@ private[ml] object FeedForwardTopology {
-    deltas(L - 1) = newE
     for (i <- (L - 2) to (0, -1)) {
-      deltas(i) = layerModels(i + 1).prevDelta(deltas(i + 1), outputs(i + 1))
+       layerModels(i + 1).prevDelta(deltas(i + 1), outputs(i + 1), deltas(i))
```
extra leading whitespace

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -226,44 +236,21 @@ private[ann] trait ActivationFunction extends Serializable {
+ * Implements in-place application of functions in the arrays
  */
-private[ann] object ActivationFunction {
+private[ann] object UniversalFunction {
```
I found it slightly confusing that the `apply` method of a companion object is mutating in place since companion object `apply`s are typically used to implement default constructors; we should consider renaming `apply` -> `call` if you think that's clearer

mengxr replied:
+1 on @feynmanliang 's suggestion. Btw, it is not clear to me what `UniversalFunction` means here. If we call it `ApplyInPlace`, then we can keep the `apply` method name.

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -466,6 +352,15 @@ private[ann] trait Topology extends Serializable{
  */
 private[ann] trait TopologyModel extends Serializable{
```
nit: Space before `{`

feynmanliang commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -532,8 +427,13 @@ private[ml] object FeedForwardTopology {
-        if (softmax && i == layerSizes.length - 2) {
-          new FunctionalLayer(new SoftmaxFunction())
+        if (i == layerSizes.length - 2) {
+          if (softmax) {
```
nit: maybe instead of `softmax`, have an `isBinaryClassification` flag to make intention clear

feynmanliang commented:
Made a pass

avulanov committed:
Addressing reviewers comments

avulanov commented:
@feynmanliang Thank you! I've addressed all comments, except related to the `ones` vector because it actually depends on the batch size. With regards to `UniversalFunction.apply`, I could call it explicitly. Does it sounds reasonable to you?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44188/consoleFull)** for PR 9229 at commit [`7864b8f`](https://github.com/apache/spark/commit/7864b8f718e565d7b4d906e004444b0345057752).

SparkQA commented:
**[Test build #44188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44188/consoleFull)** for PR 9229 at commit [`7864b8f`](https://github.com/apache/spark/commit/7864b8f718e565d7b4d906e004444b0345057752).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44188/
Test FAILed.

avulanov committed:
Fix scalastyle

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44190 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44190/consoleFull)** for PR 9229 at commit [`215d249`](https://github.com/apache/spark/commit/215d2495d2ef5af66e2345f7671edf7871352410).

SparkQA commented:
**[Test build #44190 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44190/consoleFull)** for PR 9229 at commit [`215d249`](https://github.com/apache/spark/commit/215d2495d2ef5af66e2345f7671edf7871352410).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44190/
Test FAILed.

avulanov committed:
Fix Mima missing class problem: SoftmaxFunction

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44193 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44193/consoleFull)** for PR 9229 at commit [`6f1200c`](https://github.com/apache/spark/commit/6f1200c7cf25e491a238b4e627f2d60299537014).

SparkQA commented:
**[Test build #44193 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44193/consoleFull)** for PR 9229 at commit [`6f1200c`](https://github.com/apache/spark/commit/6f1200c7cf25e491a238b4e627f2d60299537014).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44193/
Test PASSed.

avulanov committed:
Implement set weights and its unit test. Fix set seed

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45311 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45311/consoleFull)** for PR 9229 at commit [`1d29901`](https://github.com/apache/spark/commit/1d29901ef8f4cc068f12936bf41488712135656e).

SparkQA commented:
**[Test build #45311 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45311/consoleFull)** for PR 9229 at commit [`1d29901`](https://github.com/apache/spark/commit/1d29901ef8f4cc068f12936bf41488712135656e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

avulanov committed:
Fix weights bug with seed

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45410 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45410/consoleFull)** for PR 9229 at commit [`129c9aa`](https://github.com/apache/spark/commit/129c9aa50d8e3ba2a10faf8781e5ca060a6a4905).

SparkQA commented:
**[Test build #45410 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45410/consoleFull)** for PR 9229 at commit [`129c9aa`](https://github.com/apache/spark/commit/129c9aa50d8e3ba2a10faf8781e5ca060a6a4905).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

avulanov committed:
Fix seed in test to work with the new RNG

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45576/
Test FAILed.

avulanov commented:
@mengxr Could you suggest why it continues to fail? Thanks.

feynmanliang commented:
@avulanov On [this page](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45410/consoleFull) search for "*** FAILED ***"

avulanov commented:
@feynmanliang Thanks for the tip! My unit tests fail. It has a problem with the new RNG in Spark, so I updated the seeds and the asserts. However the fail seems to correspond to the old code: `Should converge to a good optimum after the 3rd restart` (**3rd**). The new code has `...after the 5th restart` (**5th**). Could you suggest why it is happening?

avulanov commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45658 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45658/consoleFull)** for PR 9229 at commit [`083a9b3`](https://github.com/apache/spark/commit/083a9b3d0f5b06ea313cd262361e8e23c4ed130e).

SparkQA commented:
**[Test build #45658 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45658/consoleFull)** for PR 9229 at commit [`083a9b3`](https://github.com/apache/spark/commit/083a9b3d0f5b06ea313cd262361e8e23c4ed130e).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45658/
Test FAILed.

avulanov committed:
Fix seed in Java MLP test

avulanov commented:
retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45680/consoleFull)** for PR 9229 at commit [`2a6d644`](https://github.com/apache/spark/commit/2a6d64412c2f19309badbb46edc40d47c0ba0d80).

SparkQA commented:
**[Test build #45680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45680/consoleFull)** for PR 9229 at commit [`2a6d644`](https://github.com/apache/spark/commit/2a6d64412c2f19309badbb46edc40d47c0ba0d80).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45680/
Test FAILed.

avulanov committed:
Fix seed in Python MLP test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45698 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45698/consoleFull)** for PR 9229 at commit [`e3ca7f1`](https://github.com/apache/spark/commit/e3ca7f1679e7ee361c9b8aaf04f4c6951521dec1).

SparkQA commented:
**[Test build #45698 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45698/consoleFull)** for PR 9229 at commit [`e3ca7f1`](https://github.com/apache/spark/commit/e3ca7f1679e7ee361c9b8aaf04f4c6951521dec1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45698/
Test PASSed.

avulanov committed:
LayerModel and TopologyModel constructor to accept weights vector

SparkQA commented:
**[Test build #45915 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45915/consoleFull)** for PR 9229 at commit [`ae3ea13`](https://github.com/apache/spark/commit/ae3ea1366d66d850e9215ae348a81203c52fd683).

SparkQA commented:
**[Test build #45915 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45915/consoleFull)** for PR 9229 at commit [`ae3ea13`](https://github.com/apache/spark/commit/ae3ea1366d66d850e9215ae348a81203c52fd683).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45915/
Test FAILed.

SparkQA commented:
**[Test build #46222 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46222/consoleFull)** for PR 9229 at commit [`d000e4e`](https://github.com/apache/spark/commit/d000e4ec833a165e77cba8b91203b11922021018).

SparkQA commented:
**[Test build #46222 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46222/consoleFull)** for PR 9229 at commit [`d000e4e`](https://github.com/apache/spark/commit/d000e4ec833a165e77cba8b91203b11922021018).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46222/
Test FAILed.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifierSuite.scala:
```diff
@@ -49,7 +49,48 @@ class MultilayerPerceptronClassifierSuite extends SparkFunSuite with MLlibTestSp
   }
 
-  // TODO: implement a more rigorous test
+  test("Test set weights by training restart") {
```
I don't quite understand this unit test. It seems that the difference is whether we want to keep using the same random number generator or reset the seed to `12L` every 10 iterations. The solution could converge to a bad local minimal but restarting doesn't always guarantee to fix it.

The content of the test doesn't match the test name. If we want to test setting initial weights, we can show that with proper warm starts, the algorithm converges faster in the same number of iterations.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/ann/GradientSuite.scala:
```diff
@@ -0,0 +1,69 @@
+      case layerWithLoss: LossFunction =>
+        layerWithLoss.loss(outputs.last, target, new BDM[Double](target.rows, target.cols))
+      case _ =>
+        throw new UnsupportedOperationException("Top layer is required to have loss.")
```
Embed layer information in the error message.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/ann/GradientSuite.scala:
```diff
@@ -0,0 +1,69 @@
+        val newModel = topology.model(Vectors.dense(weights))
+        val newLoss = computeLoss(input, target, newModel)
+        val derivativeEstimate = (newLoss - loss) / eps
+        assert((gradient(i) - derivativeEstimate) < tol)
```
* should add `math.abs`
* include the layer info in the error message. Otherwise, we don't know which one failed.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/ann/GradientSuite.scala:
```diff
@@ -0,0 +1,69 @@
+    )
+    for (layerWithError <- layersWithErrors) {
+      topology.layers(topology.layers.length - 1) = layerWithError
+      val model = topology.model(12L)
```
`seed = 12L`

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/ann/GradientSuite.scala:
```diff
@@ -0,0 +1,69 @@
+      new SigmoidLayerWithSquaredError(),
+      new SoftmaxLayerWithCrossEntropyLoss()
+    )
+    for (layerWithError <- layersWithErrors) {
```
It would be useful to comment the test method here.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/ann/GradientSuite.scala:
```diff
@@ -0,0 +1,69 @@
+    val input = new BDM[Double](3, 1, Array(1.0, 1.0, 1.0))
+    // output must contain zeros and one 1 for SoftMax
+    val target = new BDM[Double](2, 1, Array(0.0, 1.0))
+    val topology = FeedForwardTopology.multiLayerPerceptron(Array(3, 4, 2), false)
```
use named arguments on boolean variables

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/ann/GradientSuite.scala:
```diff
@@ -0,0 +1,69 @@
+
+class GradientSuite extends SparkFunSuite with MLlibTestSparkContext {
+
+  test("Gradient test") {
```
`test("gradient computation on 1-dimensional data against numerical differentiation")` (it reads better because we already have `test(...)`)

mengxr commented:
I made a quick pass on the unit tests. Will check the implementation later today.

avulanov committed:
Addressing reviewers comments

avulanov committed:
Fix since annotation

avulanov commented:
@mengxr Thank you! I did rebase and addressed your comments.

SparkQA commented:
**[Test build #52332 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52332/consoleFull)** for PR 9229 at commit [`6385112`](https://github.com/apache/spark/commit/6385112705809a2bc573ef2fe2ca7877250b78c5).

SparkQA commented:
**[Test build #52332 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52332/consoleFull)** for PR 9229 at commit [`6385112`](https://github.com/apache/spark/commit/6385112705809a2bc573ef2fe2ca7877250b78c5).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class MultilayerPerceptronClassifier @Since(\"1.5.0\") (`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52332/
Test FAILed.

avulanov committed:
Move mima exclude to 2.0.0

SparkQA commented:
**[Test build #52334 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52334/consoleFull)** for PR 9229 at commit [`3345e6f`](https://github.com/apache/spark/commit/3345e6f54c0d3cdc68c43ba41de5d5a9643eee5a).

SparkQA commented:
**[Test build #52334 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52334/consoleFull)** for PR 9229 at commit [`3345e6f`](https://github.com/apache/spark/commit/3345e6f54c0d3cdc68c43ba41de5d5a9643eee5a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52334/
Test PASSed.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -32,20 +32,43 @@ import org.apache.spark.util.random.XORShiftRandom
+   * @param inputSize input size
+   * @return output size
+   */
+  def outputSize(inputSize: Int): Int
```
`getOutputSize`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -32,20 +32,43 @@ import org.apache.spark.util.random.XORShiftRandom
    */
-  def getInstance(weights: Vector, position: Int): LayerModel
-
+  def model(weights: BDV[Double]): LayerModel
```
* `createModel` and rename `weights` to `initialWeights`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -54,39 +77,32 @@ private[ann] trait Layer extends Serializable {
    * Evaluates the data (process the data through the layer)
    * @param data data
-   * @return processed data
+   * @param output output to write to
```
document the behavior when the output matrix is not large enough

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -54,39 +77,32 @@ private[ann] trait Layer extends Serializable {
+   * @param output output to write to
    */
-  def eval(data: BDM[Double]): BDM[Double]
+  def eval(data: BDM[Double], output: BDM[Double]): Unit
```
minor: still useful to return the output instead of `Unit`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -54,39 +77,32 @@ private[ann] trait Layer extends Serializable {
-   * @param input input data
+   * @param delta delta of this layer
+   * @param output output of this layer
+   * @param pDelta storage for the result, the previous delta
```
* document what input parameters will be changed in-place
* `prevDelta`. We can change the method name to `computePrevDelta` (using a verb)

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -54,39 +77,32 @@ private[ann] trait Layer extends Serializable {
    * @return delta
    */
-  def prevDelta(nextDelta: BDM[Double], input: BDM[Double]): BDM[Double]
+  def prevDelta(delta: BDM[Double], output: BDM[Double], pDelta: BDM[Double]): Unit
```
(minor) Could return `prevDelta` instead of `Unit`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -54,39 +77,32 @@ private[ann] trait Layer extends Serializable {
    * Computes the gradient
    * @param delta delta for this layer
    * @param input input data
+   * @param cumGrad cumulative gradient
```
same here. document clearly if this going to be modified in-place

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -54,39 +77,32 @@ private[ann] trait Layer extends Serializable {
    * @param delta delta for this layer
    * @param input input data
+   * @param cumGrad cumulative gradient
    * @return gradient
```
need to update the doc

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -96,50 +112,50 @@ private[ann] trait LayerModel extends Serializable {
-    if (d == null || d.cols != nextDelta.cols) d = new BDM[Double](w.cols, nextDelta.cols)
-    BreezeUtil.dgemm(1.0, w.t, nextDelta, 0.0, d)
-    d
+  override def prevDelta(nextDelta: BDM[Double], input: BDM[Double], delta: BDM[Double]): Unit = {
```
This is not the same signature (arg names) as defined in the interface. If you use intellij, you should see a warning here.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -150,72 +166,32 @@ private[ann] object AffineLayerModel {
-    val bias = BDV.fill[Double](numOut){ (rand.nextDouble * 4.8 - 2.4) / numIn }
-    (weights, bias)
+  def randomWeights(
+    numIn: Int,
```
4-space indentation

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -150,72 +166,32 @@ private[ann] object AffineLayerModel {
+    random: Random): Unit = {
+    var i = 0
+    while (i < weights.length) {
+      weights(i) = (random.nextDouble * 4.8 - 2.4) / numIn
```
comment on `4.8` and `2.4`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -294,89 +248,13 @@ private[ann] object ActivationFunction {
-    ActivationFunction(output, target, result, m)
-    -Bsum(target :* Blog(output)) / output.cols
-  }
+  override def eval: (Double) => Double = x => 1.0 / (1 + Math.exp(-x))
```
Use `math` instead. `Math` is deprecated.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -654,12 +513,19 @@ private[ann] object FeedForwardModel {
+    for (i <- 0 until topology.layers.length) {
+      totalSize += topology.layers(i).weightSize
+    }
+    val weights = new BDV[Double](new Array[Double](totalSize))
```
`BDV.zeros`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -774,8 +640,8 @@ private[ml] class FeedForwardTrainer(
 
-  // TODO: what if we need to pass random seed?
-  private var _weights = topology.getInstance(11L).weights()
+  private var _seed = 11L
```
We use `11L` in many places. It would be nice to generate a default seed based on the class name, e.g., `this.getClass.getName.hashCode.toLong`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -783,6 +649,22 @@ private[ml] class FeedForwardTrainer(
 
   /**
+   * Returns seed
+   * @return seed
```
Not necessary to have this line.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -783,6 +649,22 @@ private[ml] class FeedForwardTrainer(
+
+  /**
+   * Sets seed
+   * @param value seed
```
Not necessary to have `@param` and `@return`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala:
```diff
@@ -783,6 +649,22 @@ private[ml] class FeedForwardTrainer(
+   * @param value seed
+   * @return trainer
+   */
+  def setSeed(value: Long): FeedForwardTrainer = {
```
`FeedForwardTrainer ` => `self.type`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+import java.util.Random
+
+import breeze.linalg.{sum => Bsum, DenseMatrix => BDM, DenseVector => BDV}
+import breeze.numerics.{log => Blog}
```
`Blog` is ambiguous. We can use `brzLog`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+   * @param output actual output
+   * @param target target output
+   * @param delta output delta to write to
+   * @return
```
* missing doc
* need to explain `delta`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+}
+
+private[ann] class SigmoidLayerWithSquaredError extends Layer {
+  override val weightSize = 0
```
leave an empty line between method declarations

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+
+private[ann] class SoftmaxLayerModelWithCrossEntropyLoss extends LayerModel with LossFunction {
+
+  val weights = new BDV[Double](0)
```
need comment

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+
+  val weights = new BDV[Double](0)
+
+  def inplaceEval(x: BDM[Double], y: BDM[Double]): Unit = {
```
* missing doc
* should it be private?
* Maybe we should merge it with `eval` because they are the same now.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+      var sum = 0.0
+      i = 0
+      while (i < x.rows) {
+        val res = Math.exp(x(i, j) - max)
```
`math`

mengxr replied:
`math`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+  override def eval(data: BDM[Double], output: BDM[Double]): Unit = {
+    inplaceEval(data, output)
+  }
+  override def prevDelta(nextDelta: BDM[Double], input: BDM[Double], delta: BDM[Double]): Unit = {}
```
Need comment on the implementation (inside `{...}`)

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala:
```diff
@@ -0,0 +1,112 @@
+  }
+  override def prevDelta(nextDelta: BDM[Double], input: BDM[Double], delta: BDM[Double]): Unit = {}
+
+  override def grad(delta: BDM[Double], input: BDM[Double], cumGrad: BDV[Double]): Unit = {}
```
ditto

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -22,7 +22,8 @@ import scala.collection.JavaConverters._
 import org.apache.spark.ml.{PredictionModel, Predictor, PredictorParams}
 import org.apache.spark.ml.ann.{FeedForwardTopology, FeedForwardTrainer}
-import org.apache.spark.ml.param.{IntArrayParam, IntParam, ParamMap, ParamValidators}
+import org.apache.spark.ml.param.{DoubleParam, IntArrayParam, IntParam, Param, ParamMap,
```
It should be okay to use `._`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -65,7 +66,42 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
 
-  setDefault(maxIter -> 100, tol -> 1e-4, layers -> Array(1, 1), blockSize -> 128)
+  /**
+   * Optimizer setup.
```
We should have the same doc as it is defined in `optimizer`.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -65,7 +66,42 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
+   * Optimizer setup.
+   * @group expertParam
+   */
+  final val optimizer: Param[String] = new Param[String](this, "optimizer",
```
We use `solver` in GLM. It might be better to be consistent.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -65,7 +66,42 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
+   * @group expertParam
+   */
+  final val optimizer: Param[String] = new Param[String](this, "optimizer",
+    " Allows setting the optimizer: minibatch gradient descent (GD) or LBFGS. " +
```
Mention which one is the default.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -65,7 +66,42 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
+  final val optimizer: Param[String] = new Param[String](this, "optimizer",
+    " Allows setting the optimizer: minibatch gradient descent (GD) or LBFGS. " +
+      " The latter is recommended one. ",
+    ParamValidators.inArray[String](Array("GD", "LBFGS")))
```
use lowercase letters: `gd` and `l-bfgs`. Note that we already used `l-bfgs` in `LinearRegression`, so we should do the same here.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -65,7 +66,42 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
+   * Learning rate.
+   * @group expertParam
+   */
+  final val learningRate: DoubleParam = new DoubleParam(this, "learning rate",
```
* use `stepSize` instead to be consistent with other algorithms under `spark.ml`
* You can get this from `HasStepSize`

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -65,7 +66,42 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
+   * Model weights. Can be returned either after training or after explicit setting
+   * @group expertParam
+   */
+  final val weights: Param[Vector] = new Param[Vector](this, "weights",
```
Is it okay to change it to `coefficients`? We did this for linear models to avoid confusion with instance weights.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -65,7 +66,42 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
+  final def getWeights: Vector = $(weights)
+
+
+  setDefault(maxIter -> 100, tol -> 1e-4, layers -> Array(1, 1), blockSize -> 128,
```
Shall we use a more reasonable default value for `layers`? Or we should leave it empty and require users to set it.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -65,7 +66,42 @@ private[ml] trait MultilayerPerceptronParams extends PredictorParams
+
+
+  setDefault(maxIter -> 100, tol -> 1e-4, layers -> Array(1, 1), blockSize -> 128,
+    optimizer -> "LBFGS", learningRate -> 0.03, weights -> null)
```
If the default is `null`, we don't need to set it.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -141,12 +178,19 @@ class MultilayerPerceptronClassifier @Since("1.5.0") (
   def setSeed(value: Long): this.type = set(seed, value)
 
+  /**
+   * Set the model weights.
```
* `Sets`
* This should be an expert param.

mengxr commented on mllib/src/main/scala/org/apache/spark/ml/classification/MultilayerPerceptronClassifier.scala:
```diff
@@ -164,11 +208,18 @@ class MultilayerPerceptronClassifier @Since("1.5.0") (
-    val mlpModel = FeedForwardTrainer.train(data)
-    new MultilayerPerceptronClassificationModel(uid, myLayers, mlpModel.weights())
+    val trainer = new FeedForwardTrainer(topology, myLayers(0), myLayers.last)
+    if ($(weights) != null) {
```
If we don't have `weights` set, this should be `isDefined(weights)`

mengxr commented:
@avulanov I made one pass on the implementation. Besides inline comments:

* It is good to keep the size checks as safe guards.
* For methods that modify the input args in-place, we need to document the behavior. Is it for both input and output, or just the output.

I'll make another pass after update.

avulanov committed:
Addressing reviewers comments

SparkQA commented:
**[Test build #53459 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53459/consoleFull)** for PR 9229 at commit [`e4e8c0d`](https://github.com/apache/spark/commit/e4e8c0d6136a1f4e578f32e72a1dd2b1abd04fd7).

SparkQA commented:
**[Test build #53461 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53461/consoleFull)** for PR 9229 at commit [`e5bef10`](https://github.com/apache/spark/commit/e5bef1080d9a62dbc238ddd19cc73c358ff429f4).

avulanov commented:
@mengxr Thank you for reviewing! I've addressed all comments, except

  - Return mutated parameter instead of `Unit`. I believe that it might be misleading because one might expect getting a new object. `Unit` gives a hint that the state of parameters might be mutated because no variable is returned.
  - Rename `weights` to `coefficients` to make it consistent with logistic regression naming. Weights is a widely accepted term in artificial neural network domain. `coefficients` might be confusing.
  - Could you elaborate on "4 space indentation"?

SparkQA commented:
**[Test build #53459 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53459/consoleFull)** for PR 9229 at commit [`e4e8c0d`](https://github.com/apache/spark/commit/e4e8c0d6136a1f4e578f32e72a1dd2b1abd04fd7).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53459/
Test PASSed.

avulanov commented:
@mengxr Could you suggest what would be the best practice to handle zeros in the computation of logarithm of output? https://github.com/avulanov/spark/blob/mlp-refactoring/mllib/src/main/scala/org/apache/spark/ml/ann/LossFunction.scala#L122

SparkQA commented:
**[Test build #53461 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53461/consoleFull)** for PR 9229 at commit [`e5bef10`](https://github.com/apache/spark/commit/e5bef1080d9a62dbc238ddd19cc73c358ff429f4).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53461/
Test FAILed.

avulanov commented:
@mengxr Rebased to accomodate https://github.com/apache/spark/pull/9854

SparkQA commented:
**[Test build #54347 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54347/consoleFull)** for PR 9229 at commit [`b21b56a`](https://github.com/apache/spark/commit/b21b56a8ab84fcb3efe1a434b9c98511b0cd6fc2).

SparkQA commented:
**[Test build #54347 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54347/consoleFull)** for PR 9229 at commit [`b21b56a`](https://github.com/apache/spark/commit/b21b56a8ab84fcb3efe1a434b9c98511b0cd6fc2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54347/
Test PASSed.

mengxr commented:
@avulanov This PR still has conflicts with master.

avulanov commented:
@mengxr That is a `project/MimaExcludes.scala`. Apparently, it is modified almost every day in the master branch. Rebased.

SparkQA commented:
**[Test build #54646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54646/consoleFull)** for PR 9229 at commit [`94dcec0`](https://github.com/apache/spark/commit/94dcec08b5bf7cb1af054e7e27b258ab0ce870a9).

SparkQA commented:
**[Test build #54646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54646/consoleFull)** for PR 9229 at commit [`94dcec0`](https://github.com/apache/spark/commit/94dcec08b5bf7cb1af054e7e27b258ab0ce870a9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54646/
Test FAILed.

avulanov commented:
Build exception unrelated to PR

mengxr commented:
test this please

SparkQA commented:
**[Test build #54671 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54671/consoleFull)** for PR 9229 at commit [`94dcec0`](https://github.com/apache/spark/commit/94dcec08b5bf7cb1af054e7e27b258ab0ce870a9).

SparkQA commented:
**[Test build #54671 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54671/consoleFull)** for PR 9229 at commit [`94dcec0`](https://github.com/apache/spark/commit/94dcec08b5bf7cb1af054e7e27b258ab0ce870a9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54671/
Test PASSed.

mengxr commented:
LGTM. Merged into master. Thanks and sorry for the long delay in code review!

avulanov commented:
@mengxr @feynmanliang Thank you!
